clickhouse/base/endpoints_iterator.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+
+#include "clickhouse/client.h"
+#include <vector>
+
+namespace clickhouse {
+
+struct ClientOptions;
+
+/**
+ * Base class for iterating through endpoints.
+*/
+class EndpointsIteratorBase
+{
+ public: 
+   virtual ~EndpointsIteratorBase() = default;
+
+   virtual void Next() = 0;
+   // Get the address of current endpoint. 
+   virtual std::string GetHostAddr() const = 0;
+
+   // Get the port of current endpoint.
+   virtual uint16_t GetPort() const = 0;
+
+   // Reset iterations.
+   virtual void ResetIterations() = 0; 
+   virtual bool NextIsExist() const = 0;
```

Comment:
```suggestion
   virtual std::optional<Endpoint> Next() = 0;    // nullopt if no more endpoints
   virtual void ResetIterations() = 0;    // start over
```

Cleaner and much simpler API

Reply:
Yes, but `ResetIterations` is needed in order not to create an infinite loop for endpoint choosing.
Otherwise we have to  check it thought `opts.endpoints.size()`,which doesn't look like a good solution. Any ideas how to do so?